"""
GUI –≤–µ—Ä—Å–∏—è —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º Qt Designer (.ui —Ñ–∞–π–ª)
–î–ª—è —Ä–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏—è –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞ –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ Qt Designer —Å —Ñ–∞–π–ª–æ–º main_window.ui
"""

import sys
import os
import subprocess
from pathlib import Path
from PyQt6 import uic
from PyQt6.QtWidgets import QApplication, QMainWindow, QFileDialog, QMessageBox
from PyQt6.QtCore import Qt
from PyQt6.QtGui import QTextCursor, QPixmap, QImage, QIcon
from gui_widgets import WorkerThread
from PyQt6.QtWidgets import QTimeEdit
from PyQt6.QtCore import QTime


class SmartTimeEdit(QTimeEdit):
    """QTimeEdit —Å –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏–º –ø–µ—Ä–µ–∫–ª—é—á–µ–Ω–∏–µ–º —Ä–∞–∑—Ä—è–¥–æ–≤ –ø—Ä–∏ –¥–æ—Å—Ç–∏–∂–µ–Ω–∏–∏ –ø—Ä–µ–¥–µ–ª–∞"""

    def stepBy(self, steps):
        """–ü–µ—Ä–µ–æ–ø—Ä–µ–¥–µ–ª—è–µ–º –ø–æ–≤–µ–¥–µ–Ω–∏–µ —Å–ø–∏–Ω–Ω–µ—Ä–æ–≤ –¥–ª—è –ø–µ—Ä–µ–∫–ª—é—á–µ–Ω–∏—è —Ä–∞–∑—Ä—è–¥–æ–≤"""
        from PyQt6.QtWidgets import QDateTimeEdit

        current = self.time()
        section = self.currentSection()

        # –ö–æ–Ω–≤–µ—Ä—Ç–∏—Ä—É–µ–º –≤ —Å–µ–∫—É–Ω–¥—ã
        total_seconds = current.hour() * 3600 + current.minute() * 60 + current.second()

        # –û–ø—Ä–µ–¥–µ–ª—è–µ–º —à–∞–≥ –≤ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –æ—Ç —Ç–µ–∫—É—â–µ–π —Å–µ–∫—Ü–∏–∏
        if section == QDateTimeEdit.Section.SecondSection:
            step_seconds = steps
        elif section == QDateTimeEdit.Section.MinuteSection:
            step_seconds = steps * 60
        elif section == QDateTimeEdit.Section.HourSection:
            step_seconds = steps * 3600
        else:
            step_seconds = steps

        # –î–æ–±–∞–≤–ª—è–µ–º —à–∞–≥
        total_seconds += step_seconds

        # –û–≥—Ä–∞–Ω–∏—á–∏–≤–∞–µ–º –º–∏–Ω–∏–º—É–º–æ–º –∏ –º–∞–∫—Å–∏–º—É–º–æ–º
        min_time = self.minimumTime()
        max_time = self.maximumTime()

        min_seconds = min_time.hour() * 3600 + min_time.minute() * 60 + min_time.second()
        max_seconds = max_time.hour() * 3600 + max_time.minute() * 60 + max_time.second()

        if total_seconds < min_seconds:
            total_seconds = min_seconds
        elif total_seconds > max_seconds:
            total_seconds = max_seconds

        # –ö–æ–Ω–≤–µ—Ä—Ç–∏—Ä—É–µ–º –æ–±—Ä–∞—Ç–Ω–æ
        hours = total_seconds // 3600
        minutes = (total_seconds % 3600) // 60
        seconds = total_seconds % 60

        self.setTime(QTime(hours, minutes, seconds))


def get_app_dir():
    """–ü–æ–ª—É—á–∏—Ç—å –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏—é –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è (—Ä–∞–±–æ—Ç–∞–µ—Ç –∏ –¥–ª—è .py –∏ –¥–ª—è .exe)"""
    if getattr(sys, "frozen", False):
        # –ó–∞–ø—É—â–µ–Ω–æ –∏–∑ exe (PyInstaller)
        return Path(sys.executable).parent
    else:
        # –ó–∞–ø—É—â–µ–Ω–æ –∏–∑ .py
        return Path(__file__).parent


def get_resource_path(relative_path):
    """–ü–æ–ª—É—á–∏—Ç—å –∞–±—Å–æ–ª—é—Ç–Ω—ã–π –ø—É—Ç—å –∫ —Ä–µ—Å—É—Ä—Å—É (UI —Ñ–∞–π–ª—ã, –∏–∫–æ–Ω–∫–∏)
    –†–∞–±–æ—Ç–∞–µ—Ç –∫–∞–∫ –¥–ª—è dev —Ä–µ–∂–∏–º–∞, —Ç–∞–∫ –∏ –¥–ª—è PyInstaller"""
    if getattr(sys, "frozen", False):
        # PyInstaller —Å–æ–∑–¥–∞–µ—Ç –≤—Ä–µ–º–µ–Ω–Ω—É—é –ø–∞–ø–∫—É _MEIPASS –¥–ª—è —Ä–µ—Å—É—Ä—Å–æ–≤
        base_path = Path(sys._MEIPASS)
    else:
        base_path = Path(__file__).parent

    return base_path / relative_path


class VideoSpellCheckerGUI(QMainWindow):
    """–ì–ª–∞–≤–Ω–æ–µ –æ–∫–Ω–æ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è –∑–∞–≥—Ä—É–∂–∞–µ–º–æ–µ –∏–∑ .ui —Ñ–∞–π–ª–∞"""

    def __init__(self):
        super().__init__()
        self.worker = None
        self.video_duration_seconds = None  # –î–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å –∑–∞–≥—Ä—É–∂–µ–Ω–Ω–æ–≥–æ –≤–∏–¥–µ–æ
        self.config_file = get_app_dir() / "config.ini"  # –ü—É—Ç—å –∫ —Ñ–∞–π–ª—É –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏

        # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –∏–∫–æ–Ω–∫—É –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è
        icon_path = get_resource_path("app.ico")
        if icon_path.exists():
            self.setWindowIcon(QIcon(str(icon_path)))

        # –ó–∞–≥—Ä—É–∂–∞–µ–º UI –∏–∑ —Ñ–∞–π–ª–∞
        ui_file = get_resource_path("main_window.ui")
        uic.loadUi(ui_file, self)

        # –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä—É–µ–º –∫–∞—Å—Ç–æ–º–Ω—ã–µ –≤–∏–¥–∂–µ—Ç—ã –¥–ª—è –≤—Ä–µ–º–µ–Ω–∏
        self.setup_time_widgets()

        # –ó–∞–≥—Ä—É–∂–∞–µ–º –Ω–∞—Å—Ç—Ä–æ–π–∫–∏
        self.load_settings()

        # –ü–æ–¥–∫–ª—é—á–∞–µ–º —Å–∏–≥–Ω–∞–ª—ã
        self.setup_connections()

        # –ó–∞–∫—Ä—ã–≤–∞–µ–º splash-—ç–∫—Ä–∞–Ω –µ—Å–ª–∏ –æ–Ω –µ—Å—Ç—å (PyInstaller)
        try:
            import pyi_splash  # type: ignore

            pyi_splash.close()
        except:
            pass

    def setup_time_widgets(self):
        """–ù–∞—Å—Ç—Ä–∞–∏–≤–∞–µ–º –≤–∏–¥–∂–µ—Ç—ã –≤—Ä–µ–º–µ–Ω–∏ (–∑–∞–º–µ–Ω—è–µ–º –Ω–∞ SmartTimeEdit)"""
        from PyQt6.QtWidgets import QTimeEdit

        # –ù–∞—Ö–æ–¥–∏–º –æ—Ä–∏–≥–∏–Ω–∞–ª—å–Ω—ã–µ QTimeEdit –≤–∏–¥–∂–µ—Ç—ã
        old_start = self.findChild(QTimeEdit, "startTimeInput")
        old_end = self.findChild(QTimeEdit, "endTimeInput")

        if not old_start or not old_end:
            print("ERROR: –ù–µ –Ω–∞–π–¥–µ–Ω—ã –≤–∏–¥–∂–µ—Ç—ã startTimeInput –∏–ª–∏ endTimeInput")
            return

        # –ü–æ–ª—É—á–∞–µ–º layout, –≤ –∫–æ—Ç–æ—Ä–æ–º –Ω–∞—Ö–æ–¥—è—Ç—Å—è –≤–∏–¥–∂–µ—Ç—ã (–æ–Ω–∏ –≤ –æ–¥–Ω–æ–º layout)
        parent_layout = old_start.parent().layout()

        # –ö–æ–ø–∏—Ä—É–µ–º –Ω–∞—Å—Ç—Ä–æ–π–∫–∏ –∏–∑ .ui —Ñ–∞–π–ª–∞
        start_settings = {
            'time': old_start.time(),
            'displayFormat': old_start.displayFormat(),
            'enabled': old_start.isEnabled(),
            'minimumSize': old_start.minimumSize(),
            'maximumSize': old_start.maximumSize(),
            'alignment': old_start.alignment(),
            'buttonSymbols': old_start.buttonSymbols()
        }

        end_settings = {
            'time': old_end.time(),
            'displayFormat': old_end.displayFormat(),
            'enabled': old_end.isEnabled(),
            'minimumSize': old_end.minimumSize(),
            'maximumSize': old_end.maximumSize(),
            'alignment': old_end.alignment(),
            'buttonSymbols': old_end.buttonSymbols()
        }

        # –°–æ–∑–¥–∞–µ–º –Ω–æ–≤—ã–µ SmartTimeEdit –≤–∏–¥–∂–µ—Ç—ã
        self.start_time_input = SmartTimeEdit()
        self.end_time_input = SmartTimeEdit()

        # –ü—Ä–∏–º–µ–Ω—è–µ–º –Ω–∞—Å—Ç—Ä–æ–π–∫–∏
        for widget, settings in [(self.start_time_input, start_settings), (self.end_time_input, end_settings)]:
            widget.setTime(settings['time'])
            widget.setDisplayFormat(settings['displayFormat'])
            widget.setEnabled(settings['enabled'])
            widget.setMinimumSize(settings['minimumSize'])
            widget.setMaximumSize(settings['maximumSize'])
            widget.setAlignment(settings['alignment'])
            widget.setButtonSymbols(settings['buttonSymbols'])

        # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º objectName –¥–ª—è –¥–æ—Å—Ç—É–ø–∞ –∫ –≤–∏–¥–∂–µ—Ç–∞–º
        self.start_time_input.setObjectName("startTimeInput")
        self.end_time_input.setObjectName("endTimeInput")

        # –í–∫–ª—é—á–∞–µ–º wrapping –¥–ª—è –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–æ–≥–æ –ø–µ—Ä–µ–∫–ª—é—á–µ–Ω–∏—è —Ä–∞–∑—Ä—è–¥–æ–≤
        self.start_time_input.setWrapping(True)
        self.end_time_input.setWrapping(True)

        # –ó–∞–º–µ–Ω—è–µ–º –≤–∏–¥–∂–µ—Ç—ã: —Å–Ω–∞—á–∞–ª–∞ —É–¥–∞–ª—è–µ–º —Å—Ç–∞—Ä—ã–µ –∏ –≤—Å—Ç–∞–≤–ª—è–µ–º –Ω–æ–≤—ã–µ
        parent_layout.replaceWidget(old_start, self.start_time_input)
        parent_layout.replaceWidget(old_end, self.end_time_input)

        # –£–¥–∞–ª—è–µ–º —Å—Ç–∞—Ä—ã–µ –≤–∏–¥–∂–µ—Ç—ã
        old_start.deleteLater()
        old_end.deleteLater()

        # –î–æ–±–∞–≤–ª—è–µ–º –ø–æ–¥—Å–∫–∞–∑–∫–∏
        self.start_time_input.setToolTip("–ù–∞—á–∞–ª—å–Ω–æ–µ –≤—Ä–µ–º—è –∞–Ω–∞–ª–∏–∑–∞ [—á–∞—Å:–º–∏–Ω:—Å–µ–∫]")
        self.end_time_input.setToolTip("–ö–æ–Ω–µ—á–Ω–æ–µ –≤—Ä–µ–º—è –∞–Ω–∞–ª–∏–∑–∞ [—á–∞—Å:–º–∏–Ω:—Å–µ–∫]")

    def setup_connections(self):
        """–ü–æ–¥–∫–ª—é—á–∞–µ–º —Å–∏–≥–Ω–∞–ª—ã –∫ —Å–ª–æ—Ç–∞–º"""
        self.browseButton.clicked.connect(self.browse_file)
        self.browseOutputButton.clicked.connect(self.browse_output_dir)
        self.startButton.clicked.connect(self.start_processing)

        # –ü–æ–¥–∫–ª—é—á–∞–µ–º –≤–∞–ª–∏–¥–∞—Ü–∏—é –≤—Ä–µ–º–µ–Ω–∏ (–≤–∏–∑—É–∞–ª—å–Ω–∞—è –∏–Ω–¥–∏–∫–∞—Ü–∏—è)
        self.start_time_input.timeChanged.connect(self.update_start_time_validation)
        self.end_time_input.timeChanged.connect(self.update_end_time_validation)

        # –°–æ—Ö—Ä–∞–Ω—è–µ–º –Ω–∞—Å—Ç—Ä–æ–π–∫–∏ –ø—Ä–∏ –∏–∑–º–µ–Ω–µ–Ω–∏–∏ –ø–∞–ø–∫–∏ –≤—ã–≤–æ–¥–∞
        self.outputDirInput.textChanged.connect(self.save_settings)

        # –ü–æ–¥–∫–ª—é—á–∞–µ–º –¥–µ–π—Å—Ç–≤–∏—è –º–µ–Ω—é
        self.actionOpenDictionary.triggered.connect(self.open_custom_dictionary)
        self.actionExit.triggered.connect(self.close)
        self.actionAbout.triggered.connect(self.show_about_dialog)

    def update_start_time_validation(self):
        """–ü—Ä–æ–≤–µ—Ä—è–µ—Ç —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–Ω–æ–µ –≤—Ä–µ–º—è –Ω–∞—á–∞–ª–∞, –æ–±–Ω–æ–≤–ª—è–µ—Ç –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏–µ –¥–ª—è –∫–æ–Ω–µ—á–Ω–æ–≥–æ –≤—Ä–µ–º–µ–Ω–∏"""

        start_time = self.start_time_input.time()
        end_time = self.end_time_input.time()
        max_end_time = self.end_time_input.maximumTime()

        self.end_time_input.blockSignals(True)

        # –í—ã—á–∏—Å–ª—è–µ–º –º–∏–Ω–∏–º–∞–ª—å–Ω–æ –¥–æ–ø—É—Å—Ç–∏–º–æ–µ –∑–Ω–∞—á–µ–Ω–∏–µ –¥–ª—è end_time
        min_end_time = start_time.addSecs(1)

        # –ï—Å–ª–∏ –º–∏–Ω–∏–º—É–º –ø—Ä–µ–≤—ã—à–∞–µ—Ç –º–∞–∫—Å–∏–º—É–º, –∏—Å–ø–æ–ª—å–∑—É–µ–º –º–∞–∫—Å–∏–º—É–º
        if min_end_time > max_end_time:
            min_end_time = max_end_time

        # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –Ω–æ–≤—ã–π –º–∏–Ω–∏–º—É–º
        self.end_time_input.setMinimumTime(min_end_time)

        # –ï—Å–ª–∏ —Ç–µ–∫—É—â–µ–µ –∑–Ω–∞—á–µ–Ω–∏–µ –º–µ–Ω—å—à–µ –Ω–æ–≤–æ–≥–æ –º–∏–Ω–∏–º—É–º–∞, –∫–æ—Ä—Ä–µ–∫—Ç–∏—Ä—É–µ–º
        if end_time < min_end_time:
            self.end_time_input.setTime(min_end_time)

        self.end_time_input.blockSignals(False)

    def update_end_time_validation(self):
        """–ü—Ä–æ–≤–µ—Ä—è–µ—Ç –∏ –∫–æ—Ä—Ä–µ–∫—Ç–∏—Ä—É–µ—Ç –≤—Ä–µ–º—è –µ—Å–ª–∏ –æ–Ω–æ –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ–µ"""
        pass

    def load_settings(self):
        """–ó–∞–≥—Ä—É–∂–∞–µ—Ç —Å–æ—Ö—Ä–∞–Ω–µ–Ω–Ω—ã–µ –Ω–∞—Å—Ç—Ä–æ–π–∫–∏"""
        from PyQt6.QtCore import QSettings

        settings = QSettings(str(self.config_file), QSettings.Format.IniFormat)

        # –ó–∞–≥—Ä—É–∂–∞–µ–º –ø–∞–ø–∫—É –≤—ã–≤–æ–¥–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: <–∫–∞—Ç–∞–ª–æ–≥ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è>/out)
        default_output_dir = str(get_app_dir() / "out")
        output_dir = settings.value("output_dir", default_output_dir)
        self.outputDirInput.setText(output_dir)

    def save_settings(self):
        """–°–æ—Ö—Ä–∞–Ω—è–µ—Ç –Ω–∞—Å—Ç—Ä–æ–π–∫–∏"""
        from PyQt6.QtCore import QSettings

        settings = QSettings(str(self.config_file), QSettings.Format.IniFormat)
        settings.setValue("output_dir", self.outputDirInput.text())
        settings.sync()

    def browse_output_dir(self):
        """–û—Ç–∫—Ä—ã—Ç—å –¥–∏–∞–ª–æ–≥ –≤—ã–±–æ—Ä–∞ –ø–∞–ø–∫–∏ –¥–ª—è —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤"""
        current_dir = self.outputDirInput.text()
        if not current_dir:
            current_dir = str(get_app_dir() / "out")

        dir_name = QFileDialog.getExistingDirectory(
            self,
            "–í—ã–±–µ—Ä–∏—Ç–µ –ø–∞–ø–∫—É –¥–ª—è —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤",
            current_dir,
            QFileDialog.Option.ShowDirsOnly
        )

        if dir_name:
            self.outputDirInput.setText(dir_name)

    def browse_file(self):
        """–û—Ç–∫—Ä—ã—Ç—å –¥–∏–∞–ª–æ–≥ –≤—ã–±–æ—Ä–∞ —Ñ–∞–π–ª–∞"""
        file_name, _ = QFileDialog.getOpenFileName(
            self,
            "–í—ã–±–µ—Ä–∏—Ç–µ –≤–∏–¥–µ–æ —Ñ–∞–π–ª",
            "",
            "Video Files (*.mp4 *.avi *.mov *.mkv *.flv *.wmv);;All Files (*.*)",
        )

        if file_name:
            self.fileInput.setText(file_name)

            # –°–±—Ä–∞—Å—ã–≤–∞–µ–º –ø—Ä–æ–≥—Ä–µ—Å—Å –∏ –æ—á–∏—â–∞–µ–º –ª–æ–≥
            self.progressBar.setValue(0)
            self.logOutput.clear()

            # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –∫–æ–Ω–µ—á–Ω–æ–µ –≤—Ä–µ–º—è –Ω–∞ –¥–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å –≤–∏–¥–µ–æ
            self.set_end_time_from_video(file_name)
            self.start_time_input.setEnabled(True)
            self.end_time_input.setEnabled(True)

    def set_end_time_from_video(self, video_path):
        """–£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ—Ç –∫–æ–Ω–µ—á–Ω–æ–µ –≤—Ä–µ–º—è —Ä–∞–≤–Ω—ã–º –¥–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏ –≤–∏–¥–µ–æ"""
        try:
            import cv2
            from PyQt6.QtCore import QTime

            cap = cv2.VideoCapture(video_path)
            if not cap.isOpened():
                return

            fps = cap.get(cv2.CAP_PROP_FPS)
            total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
            cap.release()

            if fps > 0 and total_frames > 0:
                duration_seconds = int(total_frames / fps)

                # –°–æ—Ö—Ä–∞–Ω—è–µ–º –¥–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å –≤–∏–¥–µ–æ
                self.video_duration_seconds = duration_seconds

                hours = duration_seconds // 3600
                minutes = (duration_seconds % 3600) // 60
                seconds = duration_seconds % 60

                # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –≤—Ä–µ–º—è –∫–æ–Ω—Ü–∞ –Ω–∞ –¥–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å –≤–∏–¥–µ–æ
                end_time = QTime(hours, minutes, seconds)
                self.end_time_input.setTime(end_time)

                self.start_time_input.setMaximumTime(end_time.addSecs(-1))
                self.end_time_input.setMaximumTime(end_time)

        except Exception as e:
            print(f"–ù–µ —É–¥–∞–ª–æ—Å—å –æ–ø—Ä–µ–¥–µ–ª–∏—Ç—å –¥–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å –≤–∏–¥–µ–æ: {e}")

    def check_dictionaries(self):
        """–ü—Ä–æ–≤–µ—Ä—è–µ—Ç –Ω–∞–ª–∏—á–∏–µ —Å–ª–æ–≤–∞—Ä–µ–π –∏ –ø—Ä–µ–¥–ª–∞–≥–∞–µ—Ç —Å–∫–∞—á–∞—Ç—å –ø—Ä–∏ –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ—Å—Ç–∏"""
        app_dir = get_app_dir()
        dict_dir = app_dir / "dictionaries"
        ru_aff = dict_dir / "ru_RU.aff"
        ru_dic = dict_dir / "ru_RU.dic"
        en_aff = dict_dir / "en_US.aff"
        en_dic = dict_dir / "en_US.dic"

        missing_dicts = []
        if not (os.path.exists(ru_aff) and os.path.exists(ru_dic)):
            missing_dicts.append("–†—É—Å—Å–∫–∏–π (ru_RU)")
        if not (os.path.exists(en_aff) and os.path.exists(en_dic)):
            missing_dicts.append("–ê–Ω–≥–ª–∏–π—Å–∫–∏–π (en_US)")

        if missing_dicts:
            msg = f"–ù–µ –Ω–∞–π–¥–µ–Ω—ã —Å–ª–æ–≤–∞—Ä–∏:\n" + "\n".join(f"‚Ä¢ {d}" for d in missing_dicts)
            msg += "\n\n–°–∫–∞—á–∞—Ç—å —Å–ª–æ–≤–∞—Ä–∏ –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏?"

            reply = QMessageBox.question(
                self,
                "–°–ª–æ–≤–∞—Ä–∏ –Ω–µ –Ω–∞–π–¥–µ–Ω—ã",
                msg,
                QMessageBox.StandardButton.Yes | QMessageBox.StandardButton.No,
                QMessageBox.StandardButton.Yes,
            )

            if reply == QMessageBox.StandardButton.Yes:
                return self.download_dictionaries()
            else:
                QMessageBox.information(
                    self,
                    "–ò–Ω—Ñ–æ—Ä–º–∞—Ü–∏—è",
                    "–û–±—Ä–∞–±–æ—Ç–∫–∞ –ø—Ä–æ–¥–æ–ª–∂–∏—Ç—Å—è –±–µ–∑ –ø—Ä–æ–≤–µ—Ä–∫–∏ –æ—Ä—Ñ–æ–≥—Ä–∞—Ñ–∏–∏.\n"
                    "–î–ª—è –∑–∞–≥—Ä—É–∑–∫–∏ —Å–ª–æ–≤–∞—Ä–µ–π –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ download_dictionaries.py",
                )
                return True

        return True

    def download_dictionaries(self):
        """–ó–∞–≥—Ä—É–∂–∞–µ—Ç —Å–ª–æ–≤–∞—Ä–∏"""
        import urllib.request

        try:
            # –°–æ–∑–¥–∞–µ–º –ø–∞–ø–∫—É –¥–ª—è —Å–ª–æ–≤–∞—Ä–µ–π
            app_dir = get_app_dir()
            dict_dir = app_dir / "dictionaries"
            dict_dir.mkdir(exist_ok=True)

            # URLs –¥–ª—è —Å–ª–æ–≤–∞—Ä–µ–π (–∏—Å–ø–æ–ª—å–∑—É–µ–º —Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π —Å –∞–∫—Ç—É–∞–ª—å–Ω—ã–º–∏ —Å–ª–æ–≤–∞—Ä—è–º–∏)
            dictionaries_to_download = [
                {
                    "name": "–†—É—Å—Å–∫–∏–π (ru_RU)",
                    "files": [
                        (
                            "https://raw.githubusercontent.com/LibreOffice/dictionaries/master/ru_RU/ru_RU.aff",
                            dict_dir / "ru_RU.aff",
                        ),
                        (
                            "https://raw.githubusercontent.com/LibreOffice/dictionaries/master/ru_RU/ru_RU.dic",
                            dict_dir / "ru_RU.dic",
                        ),
                    ],
                },
                {
                    "name": "–ê–Ω–≥–ª–∏–π—Å–∫–∏–π (en_US)",
                    "files": [
                        (
                            "https://raw.githubusercontent.com/LibreOffice/dictionaries/master/en/en_US.aff",
                            dict_dir / "en_US.aff",
                        ),
                        (
                            "https://raw.githubusercontent.com/LibreOffice/dictionaries/master/en/en_US.dic",
                            dict_dir / "en_US.dic",
                        ),
                    ],
                },
            ]

            # –ó–∞–≥—Ä—É–∂–∞–µ–º —Å–ª–æ–≤–∞—Ä–∏
            all_ok = True
            for dict_info in dictionaries_to_download:
                for url, dest in dict_info["files"]:
                    try:
                        urllib.request.urlretrieve(url, str(dest))
                        if not (dest.exists() and dest.stat().st_size > 100):
                            all_ok = False
                    except Exception as e:
                        all_ok = False
                        break

            if all_ok:
                QMessageBox.information(self, "–£—Å–ø–µ—Ö", "–°–ª–æ–≤–∞—Ä–∏ —É—Å–ø–µ—à–Ω–æ –∑–∞–≥—Ä—É–∂–µ–Ω—ã!")
                return True
            else:
                QMessageBox.warning(
                    self,
                    "–û—à–∏–±–∫–∞",
                    "–û—à–∏–±–∫–∞ –ø—Ä–∏ –∑–∞–≥—Ä—É–∑–∫–µ —Å–ª–æ–≤–∞—Ä–µ–π.\n\n"
                    "–ü—Ä–æ–≤–µ—Ä—å—Ç–µ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ –∏–Ω—Ç–µ—Ä–Ω–µ—Ç—É –∏–ª–∏ –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ download_dictionaries.py –≤—Ä—É—á–Ω—É—é",
                )
                return False

        except Exception as e:
            QMessageBox.warning(
                self,
                "–û—à–∏–±–∫–∞",
                f"–ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å —Å–ª–æ–≤–∞—Ä–∏:\n{str(e)}\n\n"
                "–ò—Å–ø–æ–ª—å–∑—É–π—Ç–µ download_dictionaries.py –≤—Ä—É—á–Ω—É—é",
            )
            return False

    def start_processing(self):
        """–ù–∞—á–∞—Ç—å –æ–±—Ä–∞–±–æ—Ç–∫—É –≤–∏–¥–µ–æ"""
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º –Ω–∞–ª–∏—á–∏–µ —Å–ª–æ–≤–∞—Ä–µ–π
        if not self.check_dictionaries():
            return

        video_path = self.fileInput.text()

        if not video_path:
            QMessageBox.warning(self, "–û—à–∏–±–∫–∞", "–ü–æ–∂–∞–ª—É–π—Å—Ç–∞, –≤—ã–±–µ—Ä–∏—Ç–µ –≤–∏–¥–µ–æ —Ñ–∞–π–ª")
            return

        if not os.path.exists(video_path):
            QMessageBox.warning(self, "–û—à–∏–±–∫–∞", "–í—ã–±—Ä–∞–Ω–Ω—ã–π —Ñ–∞–π–ª –Ω–µ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç")
            return

        interval = self.intervalInput.value()

        # –ü–æ–ª—É—á–∞–µ–º –≤—Ä–µ–º—è –∏–∑ QTimeEdit –∏ –∫–æ–Ω–≤–µ—Ä—Ç–∏—Ä—É–µ–º –≤ —Å–µ–∫—É–Ω–¥—ã
        from PyQt6.QtCore import QTime

        start_time = self.start_time_input.time()
        start_time_seconds = (
            start_time.hour() * 3600 + start_time.minute() * 60 + start_time.second()
        )

        end_time = self.end_time_input.time()
        # –ï—Å–ª–∏ –≤—Ä–µ–º—è 00:00:00, –∑–Ω–∞—á–∏—Ç –¥–æ –∫–æ–Ω—Ü–∞
        if end_time == QTime(0, 0, 0):
            end_time_seconds = None
        else:
            end_time_seconds = (
                end_time.hour() * 3600 + end_time.minute() * 60 + end_time.second()
            )

        # –í–∞–ª–∏–¥–∞—Ü–∏—è –≤—Ä–µ–º–µ–Ω–∏
        if end_time_seconds is not None and end_time_seconds <= start_time_seconds:
            QMessageBox.warning(
                self, "–û—à–∏–±–∫–∞", "–ö–æ–Ω–µ—á–Ω–æ–µ –≤—Ä–µ–º—è –¥–æ–ª–∂–Ω–æ –±—ã—Ç—å –±–æ–ª—å—à–µ –Ω–∞—á–∞–ª—å–Ω–æ–≥–æ"
            )
            return

        # –û—á–∏—â–∞–µ–º –ª–æ–≥ –∏ —Å–±—Ä–∞—Å—ã–≤–∞–µ–º –ø—Ä–æ–≥—Ä–µ—Å—Å
        self.logOutput.clear()
        self.progressBar.setValue(0)

        # –ë–ª–æ–∫–∏—Ä—É–µ–º –∫–Ω–æ–ø–∫–∏ –∏ –ø–æ–ª—è –≤–≤–æ–¥–∞
        self.startButton.setEnabled(False)
        self.browseButton.setEnabled(False)
        self.browseOutputButton.setEnabled(False)
        self.outputDirInput.setEnabled(False)
        self.intervalInput.setEnabled(False)
        self.start_time_input.setEnabled(False)
        self.end_time_input.setEnabled(False)
        self.startButton.setText("–û–±—Ä–∞–±–æ—Ç–∫–∞...")

        # –ü–æ–ª—É—á–∞–µ–º –±–∞–∑–æ–≤—É—é –ø–∞–ø–∫—É –≤—ã–≤–æ–¥–∞ –∏–∑ –Ω–∞—Å—Ç—Ä–æ–µ–∫
        base_output_dir = self.outputDirInput.text()
        if not base_output_dir:
            base_output_dir = str(get_app_dir() / "out")

        # –°–æ–∑–¥–∞–µ–º –ø–æ–¥–ø–∞–ø–∫—É —Å –∏–º–µ–Ω–µ–º –≤–∏–¥–µ–æ
        video_name = Path(video_path).stem
        output_dir = Path(base_output_dir) / f"screenshots_{video_name}_errors"
        output_dir = str(output_dir)

        # –°–æ–∑–¥–∞–µ–º –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏—é –µ—Å–ª–∏ –Ω–µ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç
        Path(output_dir).mkdir(parents=True, exist_ok=True)

        # –ó–∞–ø—É—Å–∫–∞–µ–º –æ–±—Ä–∞–±–æ—Ç–∫—É –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º –ø–æ—Ç–æ–∫–µ
        self.worker = WorkerThread(
            video_path, interval, output_dir, start_time_seconds, end_time_seconds
        )
        self.worker.log_signal.connect(self.append_log)
        self.worker.frame_signal.connect(self.update_frame_preview)
        self.worker.progress_signal.connect(self.update_progress)
        self.worker.finished_signal.connect(self.processing_finished)
        self.worker.error_signal.connect(self.processing_error)
        self.worker.start()

    def append_log(self, message):
        """–î–æ–±–∞–≤–∏—Ç—å —Å–æ–æ–±—â–µ–Ω–∏–µ –≤ –ª–æ–≥"""
        self.logOutput.append(message)
        self.logOutput.moveCursor(QTextCursor.MoveOperation.End)

    def update_frame_preview(self, frame):
        """–û–±–Ω–æ–≤–∏—Ç—å –ø—Ä–µ–≤—å—é —Ç–µ–∫—É—â–µ–≥–æ –∫–∞–¥—Ä–∞"""
        import cv2

        if frame is None:
            return

        # –ö–æ–Ω–≤–µ—Ä—Ç–∏—Ä—É–µ–º BGR –≤ RGB
        frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)

        # –ü–æ–ª—É—á–∞–µ–º —Ä–∞–∑–º–µ—Ä—ã
        height, width, _ = frame_rgb.shape
        bytes_per_line = 3 * width

        # –°–æ–∑–¥–∞–µ–º QImage
        q_image = QImage(
            frame_rgb.data, width, height, bytes_per_line, QImage.Format.Format_RGB888
        )

        # –ú–∞—Å—à—Ç–∞–±–∏—Ä—É–µ–º –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ —Å —Å–æ—Ö—Ä–∞–Ω–µ–Ω–∏–µ–º –ø—Ä–æ–ø–æ—Ä—Ü–∏–π
        pixmap = QPixmap.fromImage(q_image)
        scaled_pixmap = pixmap.scaled(
            self.framePreview.size(),
            Qt.AspectRatioMode.KeepAspectRatio,
            Qt.TransformationMode.SmoothTransformation,
        )

        # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –≤ label
        self.framePreview.setPixmap(scaled_pixmap)

    def update_progress(self, value):
        """–û–±–Ω–æ–≤–∏—Ç—å –ø—Ä–æ–≥—Ä–µ—Å—Å-–±–∞—Ä"""
        self.progressBar.setValue(value)

    def open_custom_dictionary(self):
        """–û—Ç–∫—Ä—ã—Ç—å —Ñ–∞–π–ª –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ —Å–ª–æ–≤–∞—Ä—è"""
        import subprocess
        from PyQt6.QtWidgets import QMessageBox

        # –ü—É—Ç—å –∫ —Ñ–∞–π–ª—É —Å–ª–æ–≤–∞—Ä—è
        app_dir = get_app_dir()
        dict_file = app_dir / "custom_dictionary.txt"

        # –°–æ–∑–¥–∞—ë–º —Ñ–∞–π–ª, –µ—Å–ª–∏ –µ–≥–æ –Ω–µ—Ç
        if not dict_file.exists():
            try:
                with open(dict_file, "w", encoding="utf-8") as f:
                    f.write("# –ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–π —Å–ª–æ–≤–∞—Ä—å\n")
                    f.write("# –î–æ–±–∞–≤—å—Ç–µ —Å–≤–æ–∏ —Å–ª–æ–≤–∞ (–ø–æ –æ–¥–Ω–æ–º—É –Ω–∞ —Å—Ç—Ä–æ–∫—É)\n")
                    f.write("# –°—Ç—Ä–æ–∫–∏, –Ω–∞—á–∏–Ω–∞—é—â–∏–µ—Å—è —Å #, –∏–≥–Ω–æ—Ä–∏—Ä—É—é—Ç—Å—è\n\n")

                QMessageBox.information(
                    self,
                    "–°–ª–æ–≤–∞—Ä—å —Å–æ–∑–¥–∞–Ω",
                    f"–°–æ–∑–¥–∞–Ω –Ω–æ–≤—ã–π —Ñ–∞–π–ª —Å–ª–æ–≤–∞—Ä—è:\n{dict_file}\n\n–î–æ–±–∞–≤—å—Ç–µ –≤ –Ω–µ–≥–æ —Å–≤–æ–∏ —Å–ª–æ–≤–∞-–∏—Å–∫–ª—é—á–µ–Ω–∏—è.",
                )
            except Exception as e:
                QMessageBox.critical(
                    self, "–û—à–∏–±–∫–∞", f"–ù–µ —É–¥–∞–ª–æ—Å—å —Å–æ–∑–¥–∞—Ç—å —Ñ–∞–π–ª —Å–ª–æ–≤–∞—Ä—è:\n{e}"
                )
                return

        # –û—Ç–∫—Ä—ã–≤–∞–µ–º —Ñ–∞–π–ª –≤ —Å–∏—Å—Ç–µ–º–Ω–æ–º —Ä–µ–¥–∞–∫—Ç–æ—Ä–µ
        try:
            if os.name == "nt":  # Windows
                os.startfile(dict_file)
            elif os.name == "posix":  # Linux/Mac
                if (
                    subprocess.call(["which", "xdg-open"], stdout=subprocess.DEVNULL)
                    == 0
                ):
                    subprocess.call(["xdg-open", dict_file])
                else:
                    subprocess.call(["open", dict_file])  # macOS
        except Exception as e:
            QMessageBox.critical(
                self,
                "–û—à–∏–±–∫–∞",
                f"–ù–µ —É–¥–∞–ª–æ—Å—å –æ—Ç–∫—Ä—ã—Ç—å —Ñ–∞–π–ª:\n{e}\n\n–ü—É—Ç—å –∫ —Ñ–∞–π–ª—É:\n{dict_file}",
            )

    def show_about_dialog(self):
        """–ü–æ–∫–∞–∑–∞—Ç—å –¥–∏–∞–ª–æ–≥ '–û –ø—Ä–æ–≥—Ä–∞–º–º–µ'"""
        from PyQt6.QtWidgets import QDialog

        # –°–æ–∑–¥–∞–µ–º –¥–∏–∞–ª–æ–≥ –∏ –∑–∞–≥—Ä—É–∂–∞–µ–º UI
        dialog = QDialog(self)
        about_ui_file = get_resource_path("about_dialog.ui")
        uic.loadUi(about_ui_file, dialog)

        # –ü–æ–∫–∞–∑—ã–≤–∞–µ–º –¥–∏–∞–ª–æ–≥
        dialog.exec()

    def processing_finished(self, result):
        """–û–±—Ä–∞–±–æ—Ç–∫–∞ –∑–∞–≤–µ—Ä—à–µ–Ω–∞ —É—Å–ø–µ—à–Ω–æ"""
        self.startButton.setEnabled(True)
        self.browseButton.setEnabled(True)
        self.browseOutputButton.setEnabled(True)
        self.outputDirInput.setEnabled(True)
        self.intervalInput.setEnabled(True)
        self.start_time_input.setEnabled(True)
        self.end_time_input.setEnabled(True)
        self.startButton.setText("–ù–∞—á–∞—Ç—å –ø—Ä–æ–≤–µ—Ä–∫—É")

        self.show_report(result)
        self.open_results_folder(result["output_dir"])

    def processing_error(self, error_message):
        """–û–±—Ä–∞–±–æ—Ç–∫–∞ –∑–∞–≤–µ—Ä—à–µ–Ω–∞ —Å –æ—à–∏–±–∫–æ–π"""
        self.startButton.setEnabled(True)
        self.browseButton.setEnabled(True)
        self.browseOutputButton.setEnabled(True)
        self.outputDirInput.setEnabled(True)
        self.intervalInput.setEnabled(True)
        self.start_time_input.setEnabled(True)
        self.end_time_input.setEnabled(True)
        self.startButton.setText("–ù–∞—á–∞—Ç—å –ø—Ä–æ–≤–µ—Ä–∫—É")

        self.append_log(f"\n‚ùå {error_message}")
        QMessageBox.critical(self, "–û—à–∏–±–∫–∞", error_message)

    def show_report(self, result):
        """–ü–æ–∫–∞–∑–∞—Ç—å –æ—Ç—á–µ—Ç –æ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞—Ö"""
        report = "\n\n" + "=" * 60 + "\n"
        report += "–û–¢–ß–ï–¢ –û –ü–†–û–í–ï–†–ö–ï\n"
        report += "=" * 60 + "\n\n"

        if result["frames_with_errors"] == 0:
            report += "‚úì –û—à–∏–±–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω–æ!\n"
        else:
            report += f"–ù–∞–π–¥–µ–Ω–æ –∫–∞–¥—Ä–æ–≤ —Å –æ—à–∏–±–∫–∞–º–∏: {result['frames_with_errors']}\n"
            report += f"–í—Å–µ–≥–æ –æ—à–∏–±–æ–∫: {result['total_errors']}\n\n"
            report += "–¢–∞–π–º-–∫–æ–¥—ã –∫–∞–¥—Ä–æ–≤ —Å –æ—à–∏–±–∫–∞–º–∏:\n"
            report += "-" * 60 + "\n"

            for detail in result["errors_details"]:
                timecode = detail["timecode"]
                minutes = int(timecode // 60)
                seconds = int(timecode % 60)
                milliseconds = int((timecode % 1) * 1000)

                report += f"\nüïê {minutes:02d}:{seconds:02d}.{milliseconds:03d} "
                report += f"(–∫–∞–¥—Ä #{detail['frame_num']}) - "
                report += f"{detail['errors_count']} –æ—à–∏–±–æ–∫\n"

                for _, error in enumerate(detail["errors"][:3]):
                    report += f"   ‚Ä¢ {error}\n"

                if len(detail["errors"]) > 3:
                    report += f"   ... –∏ –µ—â–µ {len(detail['errors']) - 3} –æ—à–∏–±–æ–∫\n"

        report += "\n" + "=" * 60 + "\n"
        self.append_log(report)

    def open_results_folder(self, folder_path):
        """–û—Ç–∫—Ä—ã—Ç—å –ø–∞–ø–∫—É —Å —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞–º–∏ –≤ –ø—Ä–æ–≤–æ–¥–Ω–∏–∫–µ"""
        if os.path.exists(folder_path):
            if sys.platform == "win32":
                os.startfile(folder_path)
            elif sys.platform == "darwin":
                subprocess.run(["open", folder_path])
            else:
                subprocess.run(["xdg-open", folder_path])


def main():
    """–ó–∞–ø—É—Å–∫ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è"""
    app = QApplication(sys.argv)
    app.setStyle("Fusion")

    # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –∏–∫–æ–Ω–∫—É –¥–ª—è –≤—Å–µ–≥–æ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è
    icon_path = get_resource_path("app.ico")
    if icon_path.exists():
        app.setWindowIcon(QIcon(str(icon_path)))

    window = VideoSpellCheckerGUI()
    window.show()

    sys.exit(app.exec())


if __name__ == "__main__":
    main()
